File: backend/app/crud/config_master.py
"""
CRUD helpers for config master data tables.
"""

from datetime import date
from typing import Any, Dict, List, Optional

from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.models.config_master import (
    CampusProfile,
    ClassProfile,
    CourseProfile,
    HomeroomTeacherProfile,
    MajorProfile,
    TeacherClassAssignment,
    TeacherProfile,
)
from app.models.human_resources import EmployeeProfile
from app.models.user import User, UserStatus
from app.schemas.config_master import (
    AssignmentCreate,
    AssignmentUpdate,
    CampusCreate,
    CampusUpdate,
    ClassCreate,
    ClassUpdate,
    CourseCreate,
    CourseUpdate,
    EmployeeCreate,
    EmployeeUpdate,
    HomeroomTeacherCreate,
    HomeroomTeacherUpdate,
    MajorCreate,
    MajorUpdate,
    TeacherCreate,
    TeacherUpdate,
)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _commit_with_refresh(db: Session, obj):
    db.add(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


# ---------------------------------------------------------------------------
# Campus
# ---------------------------------------------------------------------------


def list_campuses(db: Session, active: Optional[bool] = None) -> List[CampusProfile]:
    query = db.query(CampusProfile)
    if active is not None:
        query = query.filter(CampusProfile.is_active == active)
    return query.order_by(CampusProfile.name).all()


def get_campus(db: Session, name: str) -> Optional[CampusProfile]:
    return db.query(CampusProfile).filter_by(name=name).first()


def create_campus(db: Session, data: CampusCreate) -> CampusProfile:
    obj = CampusProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_campus(db: Session, name: str, data: CampusUpdate) -> Optional[CampusProfile]:
    obj = get_campus(db, name)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_campus(db: Session, name: str) -> bool:
    obj = get_campus(db, name)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Majors
# ---------------------------------------------------------------------------


def list_majors(
    db: Session, campus_name: Optional[str] = None, active: Optional[bool] = None
) -> List[MajorProfile]:
    query = db.query(MajorProfile)
    if campus_name:
        query = query.filter(MajorProfile.campus_name == campus_name)
    if active is not None:
        query = query.filter(MajorProfile.is_active == active)
    return query.order_by(MajorProfile.campus_name, MajorProfile.name).all()


def get_major(db: Session, major_id: int) -> Optional[MajorProfile]:
    return db.query(MajorProfile).filter_by(id=major_id).first()


def create_major(db: Session, data: MajorCreate) -> MajorProfile:
    obj = MajorProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_major(db: Session, major_id: int, data: MajorUpdate) -> Optional[MajorProfile]:
    obj = get_major(db, major_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_major(db: Session, major_id: int) -> bool:
    obj = get_major(db, major_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Courses
# ---------------------------------------------------------------------------


def list_courses(
    db: Session,
    campus_name: Optional[str] = None,
    major_id: Optional[int] = None,
    active: Optional[bool] = None,
) -> List[CourseProfile]:
    query = db.query(CourseProfile)
    if campus_name:
        query = query.filter(CourseProfile.campus_name == campus_name)
    if major_id:
        query = query.filter(CourseProfile.major_id == major_id)
    if active is not None:
        query = query.filter(CourseProfile.is_active == active)
    return query.order_by(CourseProfile.course_name).all()


def get_course(db: Session, course_id: int) -> Optional[CourseProfile]:
    return db.query(CourseProfile).filter_by(id=course_id).first()


def create_course(db: Session, data: CourseCreate) -> CourseProfile:
    obj = CourseProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_course(db: Session, course_id: int, data: CourseUpdate) -> Optional[CourseProfile]:
    obj = get_course(db, course_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_course(db: Session, course_id: int) -> bool:
    obj = get_course(db, course_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Teacher Profiles
# ---------------------------------------------------------------------------


def list_teachers(
    db: Session,
    campus_name: Optional[str] = None,
    active: Optional[bool] = None,
    participate_kpi: Optional[bool] = None,
    position: Optional[str] = None,  # 职位过滤，支持多个职位用逗号分隔
    is_primary: Optional[bool] = None,  # 是否负责强化（从teacher_class_assignments表筛选）
) -> list:
    # 如果指定了职位过滤，直接从 users 表查询
    if position:
        positions = [p.strip() for p in position.split(',') if p.strip()]
        if positions:
            # 直接从 users 表查询指定职位的用户
            query = db.query(User).filter(
                User.position.in_(positions),
                User.status == UserStatus.ACTIVE  # 只查询活跃用户
            )
            
            # 如果指定了神殿，也进行过滤
            if campus_name:
                # 处理神殿名称格式（可能带"神殿"后缀）
                campus_variants = [campus_name]
                if not campus_name.endswith('神殿'):
                    campus_variants.append(f"{campus_name}神殿")
                else:
                    campus_variants.append(campus_name.replace('神殿', ''))
                from sqlalchemy import or_
                query = query.filter(or_(*[User.campus == v for v in campus_variants]))
            
            users = query.order_by(User.real_name).all()
            
            # 将 User 对象转换为 TeacherProfile 格式的伪对象
            # 创建一个简单的类来模拟 TeacherProfile 的接口
            class TeacherProfileProxy:
                def __init__(self, user: User):
                    self.id = user.user_id
                    self.name = user.real_name
                    # 确保 campus_name 格式正确（可能需要添加"神殿"后缀）
                    campus = user.campus or ''
                    if campus and not campus.endswith('神殿'):
                        campus = f"{campus}神殿"
                    self.campus_name = campus
                    self.campus_code = user.campus or ''
                    self.user_id = user.user_id
                    self.title = user.position
                    self.phone = user.phone
                    self.email = user.email
                    self.is_active = True
                    self.participate_kpi = True
                    self.teacher_code = None
                    self.specialty = None
                    self.notes = None
                    self.created_at = user.created_at
                    self.updated_at = user.updated_at
                
                def __repr__(self):
                    return f"<TeacherProfileProxy(name={self.name}, campus={self.campus_name}, position={self.title})>"
            
            return [TeacherProfileProxy(user) for user in users]
    
    # 如果没有指定职位过滤，使用原来的逻辑
    query = db.query(TeacherProfile)
    
    if campus_name:
        query = query.filter(TeacherProfile.campus_name == campus_name)
    if active is not None:
        query = query.filter(TeacherProfile.is_active == active)
    if participate_kpi is not None:
        query = query.filter(TeacherProfile.participate_kpi == participate_kpi)
    
    # 过滤有 is_primary=true 关联记录的教员
    if is_primary is not None:
        subquery = db.query(TeacherClassAssignment.teacher_id).filter(
            TeacherClassAssignment.is_primary == is_primary
        ).distinct().subquery()
        query = query.filter(TeacherProfile.id.in_(subquery))
    
    return query.order_by(TeacherProfile.name).all()


def get_teacher(db: Session, teacher_id: int) -> Optional[TeacherProfile]:
    return db.query(TeacherProfile).filter_by(id=teacher_id).first()


def create_teacher(db: Session, data: TeacherCreate) -> TeacherProfile:
    obj = TeacherProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_teacher(db: Session, teacher_id: int, data: TeacherUpdate) -> Optional[TeacherProfile]:
    obj = get_teacher(db, teacher_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_teacher(db: Session, teacher_id: int) -> bool:
    obj = get_teacher(db, teacher_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Homeroom Teachers
# ---------------------------------------------------------------------------


def list_homeroom_teachers(
    db: Session, campus_name: Optional[str] = None, active: Optional[bool] = None
) -> List[HomeroomTeacherProfile]:
    query = db.query(HomeroomTeacherProfile)
    if campus_name:
        query = query.filter(HomeroomTeacherProfile.campus_name == campus_name)
    if active is not None:
        query = query.filter(HomeroomTeacherProfile.is_active == active)
    return query.order_by(HomeroomTeacherProfile.name).all()


def create_homeroom_teacher(
    db: Session, data: HomeroomTeacherCreate
) -> HomeroomTeacherProfile:
    payload = data.model_dump(exclude_unset=True)
    try:
        teacher = HomeroomTeacherProfile(**payload)
        return _commit_with_refresh(db, teacher)
    except Exception:
        db.rollback()
        raise


def update_homeroom_teacher(
    db: Session, homeroom_id: int, data: HomeroomTeacherUpdate
) -> Optional[HomeroomTeacherProfile]:
    teacher = db.query(HomeroomTeacherProfile).filter_by(id=homeroom_id).first()
    if not teacher:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(teacher, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(teacher)
    return teacher


def delete_homeroom_teacher(db: Session, homeroom_id: int) -> bool:
    obj = db.query(HomeroomTeacherProfile).filter_by(id=homeroom_id).first()
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Classes
# ---------------------------------------------------------------------------


def list_classes(
    db: Session,
    campus_name: Optional[str] = None,
    major_id: Optional[int] = None,
    active: Optional[bool] = None,
) -> List[Dict[str, Any]]:
    """获取班级列表，包含关联的专业、班主任、授课教员和学制信息"""
    from sqlalchemy import or_

    from app.models.teacher_employment_summary import 神殿后端教员就业汇总表
    
    query = db.query(ClassProfile)
    if campus_name:
        # 支持模糊匹配：完整名称 或 去掉"神殿"后缀的名称
        normalized_name = campus_name.rstrip('神殿').strip()
        query = query.filter(
            or_(
                ClassProfile.campus_name == campus_name,
                ClassProfile.campus_name == normalized_name,
                ClassProfile.campus_name == f"{normalized_name}神殿",
            )
        )
    if major_id:
        query = query.filter(ClassProfile.major_id == major_id)
    if active is not None:
        query = query.filter(ClassProfile.is_active == active)
    
    classes = query.order_by(ClassProfile.class_code).all()
    
    # 获取所有相关的专业和教员信息
    major_ids = [c.major_id for c in classes if c.major_id]
    homeroom_ids = [c.homeroom_teacher_id for c in classes if c.homeroom_teacher_id]
    class_ids = [c.id for c in classes]
    
    majors = {}
    if major_ids:
        major_records = db.query(MajorProfile).filter(MajorProfile.id.in_(major_ids)).all()
        majors = {m.id: m.name for m in major_records}
    
    homeroom_teachers = {}
    if homeroom_ids:
        homeroom_records = db.query(HomeroomTeacherProfile).filter(HomeroomTeacherProfile.id.in_(homeroom_ids)).all()
        homeroom_teachers = {t.id: t.name for t in homeroom_records}
    
    # 获取授课教员信息（从 TeacherClassAssignment 获取 role='instructor' 的教员）
    instructors = {}
    if class_ids:
        instructor_assignments = db.query(
            TeacherClassAssignment.class_id,
            TeacherProfile.name
        ).join(
            TeacherProfile, TeacherClassAssignment.teacher_id == TeacherProfile.id
        ).filter(
            TeacherClassAssignment.class_id.in_(class_ids),
            TeacherClassAssignment.role == 'instructor',
            TeacherClassAssignment.is_primary == True
        ).all()
        for class_id, instructor_name in instructor_assignments:
            instructors[class_id] = instructor_name
    
    # 获取学制信息（从教员就业汇总表获取）
    program_lengths = {}
    if campus_name:
        class_names = [c.class_name for c in classes]
        if class_names:
            program_records = db.query(
                神殿后端教员就业汇总表.班级名称,
                神殿后端教员就业汇总表.学制
            ).filter(
                神殿后端教员就业汇总表.神殿 == campus_name,
                神殿后端教员就业汇总表.班级名称.in_(class_names),
                神殿后端教员就业汇总表.学制 != '',
                神殿后端教员就业汇总表.学制 != None
            ).distinct().all()
            for class_name, program_length in program_records:
                if class_name not in program_lengths and program_length:
                    program_lengths[class_name] = program_length
    
    # 构建返回结果
    result = []
    for c in classes:
        # 优先使用 ClassProfile 中存储的 major_name 和 homeroom_teacher_name
        major_name = getattr(c, 'major_name', None) or (majors.get(c.major_id) if c.major_id else None)
        homeroom_name = getattr(c, 'homeroom_teacher_name', None) or (homeroom_teachers.get(c.homeroom_teacher_id) if c.homeroom_teacher_id else None)
        # 优先使用 ClassProfile 中存储的 program_length
        prog_len = getattr(c, 'program_length', None) or program_lengths.get(c.class_name)
        
        class_dict = {
            "id": c.id,
            "class_code": c.class_code,
            "class_name": c.class_name,
            "campus_name": c.campus_name,
            "major_id": c.major_id,
            "major_name": major_name,
            "homeroom_teacher_id": c.homeroom_teacher_id,
            "homeroom_teacher_name": homeroom_name,
            "instructor_name": instructors.get(c.id),  # 授课教员
            "program_length": prog_len,  # 学制
            "status": c.status,
            "start_date": c.start_date,
            "end_date": c.end_date,
            "student_capacity": c.student_capacity,
            "is_active": c.is_active,
            "notes": c.notes,
            "created_at": c.created_at,
            "updated_at": c.updated_at,
        }
        result.append(class_dict)
    
    return result


def get_class(db: Session, class_id: int) -> Optional[ClassProfile]:
    return db.query(ClassProfile).filter_by(id=class_id).first()


def create_class(db: Session, data: ClassCreate) -> ClassProfile:
    obj = ClassProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_class(db: Session, class_id: int, data: ClassUpdate) -> Optional[ClassProfile]:
    obj = get_class(db, class_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_class(db: Session, class_id: int) -> bool:
    obj = get_class(db, class_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Teacher-Class Assignments
# ---------------------------------------------------------------------------


def list_assignments(
    db: Session,
    teacher_id: Optional[int] = None,
    class_id: Optional[int] = None,
    active_only: bool = False,
) -> List[TeacherClassAssignment]:
    query = db.query(TeacherClassAssignment)
    if teacher_id:
        query = query.filter(TeacherClassAssignment.teacher_id == teacher_id)
    if class_id:
        query = query.filter(TeacherClassAssignment.class_id == class_id)
    if active_only:
        today = date.today()
        query = query.filter(
            or_(
                TeacherClassAssignment.end_date.is_(None),
                TeacherClassAssignment.end_date >= today,
            )
        )
    records = query.order_by(TeacherClassAssignment.teacher_id).all()

    # 补充 teacher_name / class_name 便于前端下拉展示（不改变表结构）
    teacher_ids = {r.teacher_id for r in records if r.teacher_id}
    class_ids = {r.class_id for r in records if r.class_id}

    teacher_name_by_id = {}
    class_name_by_id = {}
    if teacher_ids:
        for t in db.query(TeacherProfile).filter(TeacherProfile.id.in_(teacher_ids)).all():
            teacher_name_by_id[t.id] = t.name
    if class_ids:
        for c in db.query(ClassProfile).filter(ClassProfile.id.in_(class_ids)).all():
            class_name_by_id[c.id] = c.class_name

    for r in records:
        setattr(r, 'teacher_name', teacher_name_by_id.get(r.teacher_id))
        setattr(r, 'class_name', class_name_by_id.get(r.class_id))

    return records


def get_assignment(db: Session, assignment_id: int) -> Optional[TeacherClassAssignment]:
    return db.query(TeacherClassAssignment).filter_by(id=assignment_id).first()


def create_assignment(db: Session, data: AssignmentCreate) -> TeacherClassAssignment:
    obj = TeacherClassAssignment(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_assignment(
    db: Session, assignment_id: int, data: AssignmentUpdate
) -> Optional[TeacherClassAssignment]:
    obj = get_assignment(db, assignment_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_assignment(db: Session, assignment_id: int) -> bool:
    obj = get_assignment(db, assignment_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True


# ---------------------------------------------------------------------------
# Employee
# ---------------------------------------------------------------------------


def list_employees(
    db: Session,
    campus_name: Optional[str] = None,
    department: Optional[str] = None,
    position: Optional[str] = None,
    name: Optional[str] = None,
    is_active: Optional[bool] = None,
) -> List[EmployeeProfile]:
    query = db.query(EmployeeProfile)
    if campus_name:
        query = query.filter(EmployeeProfile.campus_name == campus_name)
    if department:
        query = query.filter(EmployeeProfile.department == department)
    if position:
        query = query.filter(EmployeeProfile.position == position)
    if name:
        query = query.filter(EmployeeProfile.name.ilike(f"%{name}%"))
    if is_active is not None:
        query = query.filter(EmployeeProfile.is_active == is_active)
    return query.order_by(EmployeeProfile.id).all()


def get_employee(db: Session, employee_id: int) -> Optional[EmployeeProfile]:
    return db.query(EmployeeProfile).filter_by(id=employee_id).first()


def create_employee(db: Session, data: EmployeeCreate) -> EmployeeProfile:
    obj = EmployeeProfile(**data.model_dump())
    return _commit_with_refresh(db, obj)


def update_employee(
    db: Session, employee_id: int, data: EmployeeUpdate
) -> Optional[EmployeeProfile]:
    obj = get_employee(db, employee_id)
    if not obj:
        return None
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    db.refresh(obj)
    return obj


def delete_employee(db: Session, employee_id: int) -> bool:
    obj = get_employee(db, employee_id)
    if not obj:
        return False
    db.delete(obj)
    try:
        db.commit()
    except Exception:
        db.rollback()
        raise
    return True

File: frontend/services/configMaster.ts
import type { Dayjs } from 'dayjs'
import { api } from './api'

export interface CampusProfile {
  code: string
  name: string
  campus_code?: string
  short_name?: string | null
  city?: string | null
  address?: string | null
  contact_name?: string | null
  contact_phone?: string | null
  is_active: boolean
  created_at?: string
  updated_at?: string
}

export interface MajorProfile {
  id: number
  campus_name: string  // 后端返回的是 campus_name
  name: string
  description?: string | null
  is_active: boolean
  created_at?: string
  updated_at?: string
}

export interface ClassProfile {
  id: number
  class_code?: string | null
  class_name: string
  campus_name: string  // 后端返回的是 campus_name
  campus_code?: string
  major_id?: number | null
  major_name?: string | null  // 专业名称（冗余存储）
  homeroom_teacher_id?: number | null
  homeroom_teacher_name?: string | null  // 班主任姓名（冗余存储）
  program_length?: string | null  // 学制（如：1年、1.5年、两年制）
  instructor_name?: string | null  // 授课教员名称（从关联表获取）
  status?: string | null
  start_date?: string | null
  end_date?: string | null
  student_capacity?: number | null  // 计划人数/学生人数
  is_active: boolean
  notes?: string | null
  created_at?: string
  updated_at?: string
}

export interface TeacherProfile {
  id: number
  name: string
  // 后端返回 campus_name，保留 campus_code 兼容旧调用
  campus_name?: string
  campus_code?: string
  teacher_code?: string | null
  user_id?: number | null
  title?: string | null
  phone?: string | null
  email?: string | null
  specialty?: string | null
  is_active: boolean
  participate_kpi: boolean
  notes?: string | null
  created_at?: string
  updated_at?: string
}

export interface AssignmentProfile {
  id: number
  teacher_id: number
  class_id: number
  role: string
  start_date?: string | null
  end_date?: string | null
  is_primary: boolean
  notes?: string | null
  created_at?: string
  updated_at?: string
}

export interface CourseProfile {
  id: number
  course_name: string
  campus_name: string  // 后端返回的是 campus_name
  campus_code?: string
  major_id?: number | null
  description?: string | null
  is_active: boolean
  created_at?: string
  updated_at?: string
}

export interface HomeroomTeacherProfile {
  id: number
  name: string
  campus_name: string
  campus_code?: string
  start_date?: string | null
  end_date?: string | null
  teacher_code?: string | null
  title?: string | null
  phone?: string | null
  email?: string | null
  specialty?: string | null
  is_active: boolean
  participate_kpi: boolean
  notes?: string | null
  created_at?: string
  updated_at?: string
}

export interface EmployeeProfile {
  id: number
  user_id?: number | null
  name: string
  department: string
  position: string
  contact: string
  campus_name?: string | null
  is_active: boolean
  created_at?: string
  updated_at?: string
}

const unwrap = async <T,>(promise: Promise<{ data: T }>): Promise<T> => {
  const res = await promise
  return res.data
}

// helper for serializing dates
const formatDate = (value?: Dayjs | null) => (value ? value.format('YYYY-MM-DD') : undefined)
const normalizeDateValue = (value?: string | Dayjs | null) =>
  typeof value === 'string' ? value : formatDate(value || undefined)

// campus
export const fetchCampuses = () => unwrap(api.get<CampusProfile[]>('/config/campuses'))
export const createCampus = (payload: Partial<CampusProfile>) =>
  unwrap(api.post<CampusProfile>('/config/campuses', payload))
export const updateCampus = (name: string, payload: Partial<CampusProfile>) =>
  unwrap(api.put<CampusProfile>(`/config/campuses/${encodeURIComponent(name)}`, payload))
// 注意：后端使用神殿“名称”作为路径参数
export const deleteCampus = (name: string) => unwrap(api.delete(`/config/campuses/${encodeURIComponent(name)}`))
// users by department (从 public.users 获取指定部门人员，按神殿分组)
export interface DepartmentUser {
  user_id: number
  real_name: string
  position: string | null
  phone: string | null
  campus: string
}
export interface DepartmentUserGroup {
  campus: string
  users: DepartmentUser[]
}
export const fetchUsersByDepartment = (department: string) =>
  unwrap(api.get<DepartmentUserGroup[]>('/config/department-users', { params: { department } }))
export const fetchEmployeeDepartments = () => unwrap(api.get<string[]>('/config/employees/departments'))
export const fetchEmployeePositions = () => unwrap(api.get<string[]>('/config/employees/positions'))
// majors
export const fetchMajors = (params?: { campus_code?: string; campus_name?: string; active?: boolean }) => {
  const query = {
    ...params,
    campus_name: params?.campus_name || params?.campus_code,
  }
  return unwrap(api.get<MajorProfile[]>('/config/majors', { params: query }))
}
export const createMajor = (payload: Partial<MajorProfile>) =>
  unwrap(api.post<MajorProfile>('/config/majors', payload))
export const updateMajor = (id: number, payload: Partial<MajorProfile>) =>
  unwrap(api.put<MajorProfile>(`/config/majors/${id}`, payload))
export const deleteMajor = (id: number) => unwrap(api.delete(`/config/majors/${id}`))

// teachers
export const fetchTeachers = (params?: {
  campus_code?: string
  campus_name?: string
  active?: boolean
  participate_kpi?: boolean
  position?: string  // 职位过滤，支持多个职位用逗号分隔
  is_primary?: boolean  // 是否负责强化（从teacher_class_assignments表筛选）
}) => unwrap(api.get<TeacherProfile[]>('/config/teachers', { params }))
export const createTeacher = (payload: Partial<TeacherProfile>) =>
  unwrap(api.post<TeacherProfile>('/config/teachers', payload))
export const updateTeacher = (id: number, payload: Partial<TeacherProfile>) =>
  unwrap(api.put<TeacherProfile>(`/config/teachers/${id}`, payload))
export const deleteTeacher = (id: number) => unwrap(api.delete(`/config/teachers/${id}`))

// classes
export const fetchClasses = (params?: { campus_code?: string; campus_name?: string; major_id?: number; active?: boolean }) => {
  const query = {
    ...params,
    campus_name: params?.campus_name || params?.campus_code,
  }
  return unwrap(api.get<ClassProfile[]>('/config/classes', { params: query }))
}

// 从数据库表获取唯一班级列表（用于补充配置中心的班级列表）
export const fetchClassesFromDatabase = (params?: { campus_name?: string }) =>
  unwrap(api.get<string[]>('/config/classes/from-database', { params }))
export const createClass = (payload: Partial<ClassProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.post<ClassProfile>('/config/classes', body))
}
export const updateClass = (id: number, payload: Partial<ClassProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.put<ClassProfile>(`/config/classes/${id}`, body))
}
export const deleteClass = (id: number) => unwrap(api.delete(`/config/classes/${id}`))

// assignments
export const fetchAssignments = (params?: { teacher_id?: number; class_id?: number; active_only?: boolean }) =>
  unwrap(api.get<AssignmentProfile[]>('/config/assignments', { params }))
export const createAssignment = (payload: Partial<AssignmentProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.post<AssignmentProfile>('/config/assignments', body))
}
export const updateAssignment = (id: number, payload: Partial<AssignmentProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.put<AssignmentProfile>(`/config/assignments/${id}`, body))
}
export const deleteAssignment = (id: number) => unwrap(api.delete(`/config/assignments/${id}`))

// courses
export const fetchCourses = (params?: { campus_code?: string; campus_name?: string; major_id?: number; active?: boolean }) => {
  const query = {
    ...params,
    campus_name: params?.campus_name || params?.campus_code,
  }
  return unwrap(api.get<CourseProfile[]>('/config/courses', { params: query }))
}
export const createCourse = (payload: Partial<CourseProfile>) =>
  unwrap(api.post<CourseProfile>('/config/courses', payload))
export const updateCourse = (id: number, payload: Partial<CourseProfile>) =>
  unwrap(api.put<CourseProfile>(`/config/courses/${id}`, payload))
export const deleteCourse = (id: number) => unwrap(api.delete(`/config/courses/${id}`))

// homeroom teachers
export const fetchHomeroomTeachers = (params?: { teacher_id?: number; class_id?: number; campus_name?: string; campus_code?: string; active?: boolean }) => {
  const query = {
    ...params,
    campus_name: params?.campus_name || params?.campus_code,
  }
  return unwrap(api.get<HomeroomTeacherProfile[]>('/config/homeroom-teachers', { params: query }))
}
export const createHomeroomTeacher = (payload: Partial<HomeroomTeacherProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.post<HomeroomTeacherProfile>('/config/homeroom-teachers', body))
}
export const updateHomeroomTeacher = (id: number, payload: Partial<HomeroomTeacherProfile>) => {
  const body = {
    ...payload,
    start_date: normalizeDateValue(payload.start_date),
    end_date: normalizeDateValue(payload.end_date),
  }
  return unwrap(api.put<HomeroomTeacherProfile>(`/config/homeroom-teachers/${id}`, body))
}
export const deleteHomeroomTeacher = (id: number) => unwrap(api.delete(`/config/homeroom-teachers/${id}`))

// employees (旧API，保留兼容)
export const fetchEmployees = (params?: { campus_name?: string; department?: string; position?: string; name?: string; is_active?: boolean }) => {
  return unwrap(api.get<EmployeeProfile[]>('/config/employees', { params }))
}
export const fetchDepartments = () => unwrap(api.get<string[]>('/config/employees/departments'))
export const fetchPositions = () => unwrap(api.get<string[]>('/config/employees/positions'))
export const fetchUsers = () => unwrap(api.get<any[]>('/config/employees/users'))
export const createEmployee = (payload: Partial<EmployeeProfile>) => {
  return unwrap(api.post<EmployeeProfile>('/config/employees', payload))
}
export const updateEmployee = (id: number, payload: Partial<EmployeeProfile>) => {
  return unwrap(api.put<EmployeeProfile>(`/config/employees/${id}`, payload))
}
export const deleteEmployee = (id: number) => unwrap(api.delete(`/config/employees/${id}`))
export const importEmployeesFromUsers = () => {
  return unwrap(api.post<{ success: boolean; imported: number; skipped: number; message: string }>('/config/employees/import', {}))
}

// ========== 用户权限管理 (直接操作 users 表) ==========

/** 用户权限配置信息 */
export interface UserPermissionInfo {
  user_id: number
  username: string
  name: string
  department: string | null
  position: string | null
  campus: string | null
  phone: string | null
  email: string | null
  role: string | null
  is_superuser: boolean
  status: string
}

/** 获取用户权限配置列表 */
export const fetchUserPermissions = (params?: { 
  campus?: string
  department?: string
  position?: string
  name?: string 
}) => {
  return unwrap(api.get<UserPermissionInfo[]>('/config/user-permissions', { params }))
}

/** 获取单个用户权限配置 */
export const fetchUserPermission = (userId: number) => {
  return unwrap(api.get<UserPermissionInfo>(`/config/user-permissions/${userId}`))
}

/** 更新用户权限配置 */
export const updateUserPermission = (userId: number, data: {
  department?: string
  position?: string
  campus?: string
  phone?: string
  email?: string
  status?: string
}) => {
  return unwrap(api.put<UserPermissionInfo & { success: boolean; message: string }>(
    `/config/user-permissions/${userId}`, 
    data
  ))
}

/** 创建新员工 */
export const createUserEmployee = (data: {
  username: string
  password: string
  real_name: string
  department: string
  position: string
  campus?: string
  phone?: string
  email?: string
  gender?: string
}) => {
  return unwrap(api.post<UserPermissionInfo & { success: boolean; message: string }>(
    '/config/user-permissions',
    data
  ))
}

/** 删除员工（软删除） */
export const deleteUserEmployee = (userId: number) => {
  return unwrap(api.delete<{ success: boolean; message: string }>(
    `/config/user-permissions/${userId}`
  ))
}

/** 获取部门选项 */
export const fetchDepartmentOptions = () => {
  return unwrap(api.get<string[]>('/config/user-permissions/options/departments'))
}

/** 获取职位选项 */
export const fetchPositionOptions = () => {
  return unwrap(api.get<string[]>('/config/user-permissions/options/positions'))
}

/** 获取神殿选项 */
export const fetchCampusOptions = () => {
  return unwrap(api.get<string[]>('/config/user-permissions/options/campuses'))
}
